### ISSUE TITLE
Collapse plugin lacks option to disable animation

### ISSUE THREAD

**Issue by gtristan** [twbs/bootstrap#15747]
Sometimes you just need to syncrhonously show or hide the content of a collapse.

Bootstrap currently does not provide any API to do this, I've been trying various combinations
of adding and removing the 'in' class from the collapse element without success.

Further, it seems one cannot even work around the situation by calling $(element).collapse ('destroy')
to just manually show/hide the thing and re-initialize the collapse.

**Comment by kkirsche** [comment]
Why can't you just use:

```javascript
$('.collapse').collapse('show');
```

As that will show the collapsed element then hide it again with:
```javascript
$('.collapse').collapse('hide');
```

See this:
http://jsbin.com/niyijadule/1/edit?html,js,output

**Comment by twbs-lmvtfy** [comment]
Hi @kkirsche!

You appear to have posted a live example (http://jsbin.com/niyijadule/1/edit), which is always a good first step. However, according to [Bootlint](https://github.com/twbs/bootlint), **your example has some Bootstrap usage errors**, which might potentially be causing your issue:
* [W002](https://github.com/twbs/bootlint/wiki/W002): `<head>` is missing X-UA-Compatible `<meta>` tag that disables old IE compatibility modes
* [W003](https://github.com/twbs/bootlint/wiki/W003): `<head>` is missing viewport `<meta>` tag that enables responsiveness

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by kkirsche** [comment]
Fixed: http://jsbin.com/cayoripevo/1/edit?html,js,output

**Comment by gtristan** [comment]
@kkirsche I'm not sure you understood the issue.

I am already using $(element).collapse ('show'); and $(element).collapse('hide'); to trigger the nice and fancy collapse animation.

The problem is when I need to show it right away, or hide it right away, without any animation, so that it is shown/hidden immediately.

**Comment by kkirsche** [comment]
OK, so you essentially would like the `animation` boolean that exists on tooltips, popovers, etc. added to the collapse module. Sorry, I didn't understand that from your first post.

**Comment by kkirsche** [comment]
Yep should just be checking the option
https://github.com/twbs/bootstrap/blob/master/js/collapse.js#L87
and
https://github.com/twbs/bootstrap/blob/master/js/collapse.js#L126

So the transition either occurs or not. If the admins are interested I can do a pull request for this easily.

**Comment by gtristan** [comment]
@kkirsche I'm not sure that that would do the trick.

From what I understand, the animation property on a tooltip and such is a one-off thing you decide at initialization of the tooltip or popover. That seems a bit pointless with a collapse (as you could just show or hide the content without an animation just as easily without using any collapse). Instead, what I need is a way to immediately show or hide the content of an existing collapse, as a separate semantic from the regular animated show/hide.

To be extra clear, let me explain the use case which led me to filing this.

What I have is a sort of treeview, which is a table with parent rows and child rows, the parent rows have an expander on the left some text and an "Add" button on the right (glyphicon-plus)

When the user clicks on the expander, it reveals or hides the child rows for that parent row, using the collapse animation (this part works well).

When the user clicks the "Add" button, a modal pops up, and the user enters some information to add something new to the list, as a child of that row.

What I want to do in the second case, is immediately ensure that that parent row has it's children shown, and then I do my own animation to smoothly scroll the treeview to ensure the newly added and selected item is visible.

Right now my work around is:
* trigger the collapse animation with $(children).collapse('show')
* trap the shown event with $(children).one ('shown.bs.collapse')
* when the children are finally expanded, do my other animation to ensure the new item is visible

But this double animation is clunky, also note that it is impossible to trigger both animations at once, because one needs to know the $(element).offset() of the already visible element inside the list in order to properly animate scrollTop to the correct position.

**Comment by kkirsche** [comment]
In that case you are correct but the underlying idea is the same. Yes you can and should have a default when you create it, which would be true. This is your fallback. Then you just allow it to be overriden with optional secondary argument you could disable things on the fly `$('.collapse').collapse('show', { "animation": false });` for example though I'm not sure this is an issue.

Because that call is what would allow you to avoid the transition, all you need is a boolean to decide if `complete.call(this)` should be called or not. If not, you transition and then ends. If it is, then there is no transition.

**Comment by gtristan** [comment]
Ah great, yes that would be fine, I wasnt aware that you could specify options directly in the method calls like that.

**Comment by kkirsche** [comment]
Currently you can't but all that would happen would be you would add a second argument to the method then read it in a way similar to:

```javascript
optionalArg = (typeof optionalArg === "undefined") ? "defaultValue" : optionalArg;
```

This way you either have an undefined variable if it wasn't passed in so you set the value or you use what was passed in if it wasn't.

Currently it's not possible but it's not hard to add something like that if @mdo / admins felt it was worth adding to v3 or if it was something that would wait for v4's release

**Comment by Nikita240** [comment]
+1

**Comment by mdo** [comment]
Not adding features to v3, so punting on this for now.

**Comment by wnr** [comment]
Is this coming to v4?

**Comment by cvrebert** [comment]
@wnr Transitions can be disabled in v3 (and probably also v4) via http://getbootstrap.com/javascript/#disabling-transitions

**Comment by gtristan** [comment]
@cvrebert, no, not in such a way that you can animate a single collapse sometimes, and set the widget to be collapsed or not programatically, while having the option to animate.

Please read the bug report carefully, @kkirsche originally misinterpreted this as well.

What is required, is a semantic on the open/close API of this widget, an added boolean "animate" or equivalent feature which can be used at show/hide time.

**Comment by gtristan** [comment]
fwiw, this has been a while but I did attempt some hacks.

I dont believe this will be as easy as @kkirsche suspects, there is a reason that this is not exposed in the bootstrap API, which is that boostrap relies on CSS transitions, which are inherently buggy and dont seem to support this, at least not easily. The transition holds some internal state about the HTML node, and seems to have no semantics to "execute without animating", or for clearing it's internal state so that one can show/hide the thing and re-apply the transition to the node afterwards.

I am not convinced that the required behaviour is achievable with transitions, as such I would be satisfied with a documentation update noting the caveat as a means to close this bug. If the policy is to stick with CSS transitions; this can assist at least developers in deciding if the collapse class is a good fit for their application early.

**Comment by cvrebert** [comment]
> What is required, is a semantic on the open/close API of this widget, an added boolean "animate" or equivalent feature which can be used at show/hide time.

@gtristan @wnr Ah. Feel free to file a new feature request issue for Bootstrap v4 then.

**Comment by gtristan** [comment]
@cvrebert, I took sufficient time reporting this to the bootstrap maintainership.

I have moved on of course and found a reliable workaround which is to not use the CSS transition for this use case at all, but I did invest time to be a good netizen and assist the bootstrap maintainership by highlighting this issue so that it could be addressed.

If you are to lose this information in your bug tracking, that reflects on you. I wont be held hostage by this.

**Comment by kkirsche** [comment]
I wouldn't say it's something about losing or keeping the information, I think it's a question of whether someone feels strong enough to create a new issue. 

Different versions have different implementations and there are different defaults in some cases and re-evaluating and considering what features should or shouldn't be implemented in a version. V4 removed glyph icons for examples. So the question I'd ask is do you feel strong enough about a feature like this to open a new issue and reference this thread to provide detail or do you feel you've done what you needed to and can move on comfortably?

Either way, best of luck :smiley:

**Comment by wnr** [comment]
I've created a separate issue for the v4 feature request.
